File: app/src/main/java/com/sjn/stamp/media/provider/multiple/MultipleListProvider.java
package com.sjn.stamp.media.provider.multiple;

import android.content.Context;
import android.content.res.Resources;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.MediaMetadataCompat;

import com.google.common.collect.Lists;
import com.sjn.stamp.media.provider.ListProvider;
import com.sjn.stamp.utils.LogHelper;
import com.sjn.stamp.utils.MediaIDHelper;
import com.sjn.stamp.utils.MediaItemHelper;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

abstract class MultipleListProvider extends ListProvider {

    private static final String TAG = LogHelper.makeLogTag(MultipleListProvider.class);

    protected Context mContext;
    private Map<String, List<MediaMetadataCompat>> mTrackListMap = new ConcurrentHashMap<>();

    abstract protected String getMediaKey();

    abstract protected int compareMediaList(MediaMetadataCompat lhs, MediaMetadataCompat rhs);

    MultipleListProvider(Context context) {
        mContext = context;
    }

    @Override
    final public void reset() {
        mTrackListMap.clear();
    }

    protected Map<String, List<MediaMetadataCompat>> createTrackListMap(final Map<String, MediaMetadataCompat> musicListById) {
        Map<String, List<MediaMetadataCompat>> trackListMap = new HashMap<>();

        for (MediaMetadataCompat m : musicListById.values()) {
            //noinspection ResourceType
            String key = MediaIDHelper.escape(m.getString(getMediaKey()));
            if (key != null && !key.isEmpty()) {
                List<MediaMetadataCompat> list = trackListMap.get(key);
                if (list == null) {
                    list = new ArrayList<>();
                    trackListMap.put(key, list);
                }
                list.add(m);
            }
        }
        return trackListMap;
    }

    @Override
    final public List<MediaBrowserCompat.MediaItem> getListItems(String mediaId, Resources resources, ProviderState state, final Map<String, MediaMetadataCompat> musicListById) {
        List<MediaBrowserCompat.MediaItem> items = new ArrayList<>();
        if (MediaIDHelper.isTrack(mediaId)) {
            return items;
        }
        if (getProviderMediaId().equals(mediaId)) {
            for (String key : getKeys(state, musicListById)) {
                items.add(createBrowsableMediaItemForKey(key));
            }
        } else if (mediaId.startsWith(getProviderMediaId())) {
            String key = MediaIDHelper.getHierarchy(mediaId)[1];
            for (MediaMetadataCompat item : getListByKey(key, state, musicListById)) {
                items.add(createMediaItem(item, key));
            }
        } else {
            LogHelper.w(TAG, "Skipping unmatched mediaId: ", mediaId);
        }
        return items;
    }

    /**
     * Get music tracks of the given key
     */
    @Override
    final public List<MediaMetadataCompat> getListByKey(String key, ProviderState state, final Map<String, MediaMetadataCompat> musicListById) {
        if (state != ProviderState.INITIALIZED || !getTrackListMap(musicListById).containsKey(key)) {
            return Collections.emptyList();
        }
        List<MediaMetadataCompat> list = new ArrayList<>(getTrackListMap(musicListById).get(key));
        Collections.sort(list, new Comparator<MediaMetadataCompat>() {
            @Override
            public int compare(MediaMetadataCompat lhs, MediaMetadataCompat rhs) {
                return compareMediaList(lhs, rhs);
            }
        });
        return list;
    }

    @Override
    protected MediaBrowserCompat.MediaItem createMediaItem(MediaMetadataCompat metadata, String key) {
        return MediaItemHelper.createPlayableItem(MediaItemHelper.updateMediaId(metadata, createHierarchyAwareMediaID(metadata)));
    }

    private List<String> getKeys(ProviderState state, Map<String, MediaMetadataCompat> musicListById) {
        if (state != ProviderState.INITIALIZED) {
            return Collections.emptyList();
        }
        Map<String, List<MediaMetadataCompat>> trackListMap = getTrackListMap(musicListById);
        if (trackListMap == null || trackListMap.isEmpty()) {
            return new ArrayList<>();
        }
        List<String> list = Lists.newArrayList(trackListMap.keySet());
        Collections.sort(list);
        return list;
    }

    private String createHierarchyAwareMediaID(MediaMetadataCompat metadata) {
        //noinspection ResourceType
        String category = metadata.getString(getMediaKey());
        return MediaIDHelper.createMediaID(metadata.getDescription().getMediaId(), getProviderMediaId(), category);
    }

    private MediaBrowserCompat.MediaItem createBrowsableMediaItemForKey(String key) {
        return MediaItemHelper.createBrowsableItem(MediaIDHelper.createMediaID(null, getProviderMediaId(), key), MediaIDHelper.unescape(key));
    }

    protected Map<String, List<MediaMetadataCompat>> getTrackListMap(final Map<String, MediaMetadataCompat> musicListById) {
        if (mTrackListMap == null || mTrackListMap.isEmpty()) {
            mTrackListMap = createTrackListMap(musicListById);
        }
        return mTrackListMap;
    }
}

File: app/src/main/java/com/sjn/stamp/media/provider/multiple/MultipleListProvider.kt
package com.sjn.stamp.media.provider.multiple

import android.content.Context
import android.content.res.Resources
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.MediaMetadataCompat
import com.google.common.collect.Lists
import com.sjn.stamp.media.provider.ListProvider
import com.sjn.stamp.utils.LogHelper
import com.sjn.stamp.utils.MediaIDHelper
import com.sjn.stamp.utils.MediaItemHelper
import java.util.*
import java.util.concurrent.ConcurrentHashMap

abstract class MultipleListProvider(protected var context: Context) : ListProvider() {

    protected val trackMap: MutableMap<String, MutableList<MediaMetadataCompat>> = ConcurrentHashMap()

    protected abstract val mediaKey: String

    protected abstract fun compareMediaList(lhs: MediaMetadataCompat, rhs: MediaMetadataCompat): Int

    override fun reset() = trackMap.clear()

    protected open fun updateTrackListMap(musicListById: MutableMap<String, MediaMetadataCompat>) {
        for (music in musicListById.values) {
            val key = MediaIDHelper.escape(music.getString(mediaKey) ?: "")
            if (!key.isEmpty()) {
                var list = trackMap[key]
                if (list == null) {
                    list = ArrayList()
                    trackMap[key] = list
                }
                list.add(music)
            }
        }
    }

    override fun getListItems(mediaId: String, resources: Resources, state: ListProvider.ProviderState, musicListById: MutableMap<String, MediaMetadataCompat>): MutableList<MediaBrowserCompat.MediaItem> {
        val items = ArrayList<MediaBrowserCompat.MediaItem>()
        if (MediaIDHelper.isTrack(mediaId)) {
            return items
        }
        when {
            providerMediaId == mediaId -> getKeys(state, musicListById).mapTo(items) { createBrowsableMediaItemForKey(it) }
            mediaId.startsWith(providerMediaId) -> {
                MediaIDHelper.getHierarchy(mediaId)[1].let { key ->
                    getListByKey(key, state, musicListById).mapTo(items) { createMediaItem(it, key) }
                }
            }
            else -> LogHelper.w(TAG, "Skipping unmatched mediaId: ", mediaId)
        }
        return items
    }

    /**
     * Get music tracks of the given key
     */
    override fun getListByKey(key: String?, state: ListProvider.ProviderState, musicListById: MutableMap<String, MediaMetadataCompat>): List<MediaMetadataCompat> {
        if (state !== ListProvider.ProviderState.INITIALIZED || !getTrackListMap(musicListById).containsKey(key)) {
            return emptyList()
        }
        return ArrayList(getTrackListMap(musicListById)[key]).apply {
            this.sortWith(Comparator { lhs, rhs -> compareMediaList(lhs, rhs) })
        }
    }

    override fun createMediaItem(metadata: MediaMetadataCompat, key: String): MediaBrowserCompat.MediaItem = MediaItemHelper.createPlayableItem(MediaItemHelper.updateMediaId(metadata, createHierarchyAwareMediaID(metadata)))

    private fun getKeys(state: ListProvider.ProviderState, musicListById: MutableMap<String, MediaMetadataCompat>): List<String> {
        if (state !== ListProvider.ProviderState.INITIALIZED) {
            return emptyList()
        }
        val trackListMap = getTrackListMap(musicListById)
        if (trackListMap.isEmpty()) {
            return ArrayList()
        }
        val list = Lists.newArrayList(trackListMap.keys)
        list.sort()
        return list
    }

    private fun createHierarchyAwareMediaID(metadata: MediaMetadataCompat): String = MediaIDHelper.createMediaID(metadata.description.mediaId, providerMediaId, metadata.getString(mediaKey)
            ?: "")

    private fun createBrowsableMediaItemForKey(key: String): MediaBrowserCompat.MediaItem = MediaItemHelper.createBrowsableItem(MediaIDHelper.createMediaID(null, providerMediaId, key), MediaIDHelper.unescape(key))

    protected open fun getTrackListMap(musicListById: MutableMap<String, MediaMetadataCompat>): MutableMap<String, MutableList<MediaMetadataCompat>> {
        if (trackMap.isEmpty()) {
            updateTrackListMap(musicListById)
        }
        return trackMap
    }

    companion object {

        private val TAG = LogHelper.makeLogTag(MultipleListProvider::class.java)
    }
}
